### ISSUE TITLE
Primitive arrays have a high impact on .vo file size

### ISSUE THREAD

**Issue by Janno** [rocq-prover/rocq#18646]
#### Description of the problem

I am using the following file as a template (with one of the two definitions commented out) to test the cost of primitive arrays in terms of .vo file size. The comparison is against a list containing the same values, including the array's default value as well as an extra entry at the beginning to represent the length of the array. I am including the length to be as fair as possible.

The numbers in the array have been chosen without much care except that my real world example has integers in those ranges.

```coq
From Coq Require Import List PArray Uint63.
(* Definition test := Eval lazy in List.repeat ([| 0x14%uint63; 0x61616161616161%uint63; 0x61616162626262%uint63; 0x62626262626200%uint63 | 0x0%uint63 : int |]) N. *)
Import ListNotations.
Definition test := Eval lazy in List.repeat ([ 0x05%uint63; (* length of the array *)  0x14%uint63; 0x61616161616161%uint63; 0x61616162626262%uint63; 0x62626262626200%uint63; 0x0%uint63 ]) N.
```
Replacing `N` by various numbers I arrive at the following .vo file sizes on 8.18:

<details><summary>Table</summary>

N | Array | List
-- | -- | --
1 | 21028 | 20898
2 | 21117 | 20942
3 | 21206 | 20986
4 | 21298 | 21030
5 | 21387 | 21074
6 | 21476 | 21118
7 | 21566 | 21162
8 | 21655 | 21206
9 | 21744 | 21250
10 | 21835 | 21296
20 | 22725 | 21736
30 | 23615 | 22176
40 | 24505 | 22616
50 | 25395 | 23056
60 | 26288 | 23496
70 | 27178 | 23936
80 | 28068 | 24376
90 | 28959 | 24816
100 | 29851 | 25258
150 | 34414 | 27579
200 | 39014 | 29929
250 | 43614 | 32279
300 | 48214 | 34629
350 | 52814 | 36979
450 | 62014 | 41679
500 | 66614 | 44029
600 | 75814 | 48729
700 | 85014 | 53429
800 | 94214 | 58129
900 | 103414 | 62829
1000 | 112616 | 67531
2000 | 204616 | 114531
3000 | 296616 | 161531
4000 | 388616 | 208531
5000 | 480616 | 255531
6000 | 572616 | 302531
7000 | 664616 | 349531
8000 | 756616 | 396537
9000 | 848616 | 443539
10000 | 940618 | 490541

</details>

As a graph:

![image](https://github.com/coq/coq/assets/70595/fc8469b1-f5f1-4a30-aa35-84d20c31e3f5)

I have tried to investigate this myself but I am not getting anywhere due to lack of tooling around .vo files. I think I can exclude the term representation itself since I have done the math for the expected sizes and the expected sharing and validated it against real .vo files.  I think the most likely candidate is the vm bytecode that is emitted for array literals. The bytecode format in general seems very wasteful but for the instructions used by array literals I am seeing almost 75% `0` bytes in the .vo file, I think.

#### Coq Version
8.18

**Comment by SkySkimmer** [comment]
Dump Bytecode says the list is turned into a precomputed value but the array stays as code.
If you prevent precomputing by using global names (eg `[| nat; int | bool|]` vs `[nat; int; bool]`) what do the curves look like?

**Comment by ppedrot** [comment]
Indeed, the bytecode representation on disk is not super optimized, I have tried a few different patches to compress it, but then the tradeoff is the size on disk vs. runtime complexity of compression. I can unearth these as PRs to double-check on this example, but I'm not entirely convinced it is worth it. The cost is still linear in the length of the array, and compared to the average vo size, 500ko for an array of length 10.000 doesn't seem that far-fetched.

**Comment by SkySkimmer** [comment]
We could try to recognize Array as possbly being values around https://github.com/coq/coq/blob/3d1668c41b2700cdf59310967b4cc7473a24a4f6/kernel/genlambda.ml#L716-L723

**Comment by ppedrot** [comment]
The problem here is not the bytecode per se, it's the structured value associated to the array which is stored as-is on the disk. I don't quite understand why there is a difference but we represent the value as a list in the VM, hence the overhead.

**Comment by Janno** [comment]
> Dump Bytecode says the list is turned into a precomputed value but the array stays as code. If you prevent precomputing by using global names (eg `[| nat; int | bool|]` vs `[nat; int; bool]`) what do the curves look like?

You are on to something, it seems:

![image](https://github.com/coq/coq/assets/70595/dc85a063-6fdb-4341-9a28-199dce57953c)


> Indeed, the bytecode representation on disk is not super optimized, I have tried a few different patches to compress it, but then the tradeoff is the size on disk vs. runtime complexity of compression. I can unearth these as PRs to double-check on this example, but I'm not entirely convinced it is worth it. 

@ppedrot I am very intrigued by the tradeoff of size versus runtime. I don't have experience writing bytecode interpreters so I'd love to see what you tried. AFAICT  Coq's vm instructions generally waste three bytes (see `out` in vmemitcodes.ml).  My naive approach would be to pack arguments like `tag` of the various `MAKEBLOCK`s into those bytes, if possible. If I am reading OCaml's mlvalues.h correctly the `tag` is just one byte.

I tried implementing this just now but C code is the bane of my existence and I couldn't find all the places where I have to add a mask to make it work.

> The cost is still linear in the length of the array, and compared to the average vo size, 500ko for an array of length 10.000 doesn't seem that far-fetched.

The problem is that this negates one advantage that primitive arrays arguably should have over lists: compactness on disk. I stumbled into this issue because I was doing experiments with efficient string representations. Despite my chosen representation having twice the density in the term representation, the .vo files still ended up being about 50% bigger than the usual (specialized) list of bytes with 256 constructors.

**Comment by SkySkimmer** [comment]
Here is an experiment storing the array as an array on disk: b6175452ef364c8cb26a21059be580ca5e39099c

But I'm not sure it's a good idea as explained in the commit message, because running the VM on something which mutates the array will mutate the value (replacing it by the mutated array + backtrack information).

I guess we could add an instruction Kmakeparray to replace the 3 makeblock, not sure how much that would help.

**Comment by Janno** [comment]
> But I'm not sure it's a good idea as explained in the commit message, because running the VM on something which mutates the array will mutate the value (replacing it by the mutated array + backtrack information).

Could the array not be stored as a value and be copied entirely when it is used? I don't understand the vm setup enough to say how feasible this is. Is there a "static segment" where the value could be put? If so, we'd need only a single memcopy-like instruction and a pointer to that segment and, additionally, the segment length if that would not be included in the segment itself. This would already reduced the overhead a lot, I think.

> I guess we could add an instruction Kmakeparray to replace the 3 makeblock, not sure how much that would help.

I guess my vague proposal above boils down to a more generic version of that `Kmakeparray` instruction but I don't know if the generic one would actually be useful for anything else. Arrays are really the only constrs that are mutable under the hood, aren't they?

**Comment by silene** [comment]
> Arrays are really the only constrs that are mutable under the hood, aren't they?

No, cofixpoints are mutable too.

> running the VM on something which mutates the array will mutate the value

Yes, this seems like a bad idea. And I think we had a similar bug in the past.

I guess we should just do it the same way OCaml does: store an array, but put a call to `Obj.dup` in the bytecode / native code. (Yes, even `ocamlopt` does that.) This will not solve the issue with the constant cost of creating the 3 blocks of a primitive array, but it will at least nullify the overhead of cells.

**Comment by ppedrot** [comment]
AFAIU this was solved by #19034.
